File: app/schemas/log.py
"""
日誌相關的 Pydantic 模型
"""

from typing import List, Optional
from pydantic import BaseModel, field_validator
from datetime import datetime
from app.db.models import EventLogStatus, DispatchLogStatus


class EventLogResponse(BaseModel):
    """事件日誌響應模型"""

    id: str
    topic_id: str
    source_ip: Optional[str] = None
    headers: Optional[dict] = None
    content_type: Optional[str] = None
    payload: Optional[str] = None
    status: str
    received_at: datetime

    class Config:
        from_attributes = True

    @field_validator('status', mode='before')
    @classmethod
    def status_enum_to_string(cls, v):
        """將狀態枚舉轉換為字符串"""
        if isinstance(v, EventLogStatus):
            return v.name
        return v


class DispatchLogResponse(BaseModel):
    """派發日誌響應模型"""

    id: str
    event_log_id: str
    subscription_id: str
    status: Optional[str] = None
    response_status_code: Optional[int] = None
    response_body: Optional[str] = None
    dispatched_at: datetime

    class Config:
        from_attributes = True

    @field_validator('status', mode='before')
    @classmethod
    def status_enum_to_string(cls, v):
        """將狀態枚舉轉換為字符串"""
        if isinstance(v, DispatchLogStatus):
            return v.name
        return v


class EventLogListResponse(BaseModel):
    """事件日誌列表響應模型"""

    items: List[EventLogResponse]
    total: int
    skip: int
    limit: int


class DispatchLogListResponse(BaseModel):
    """派發日誌列表響應模型"""

    items: List[DispatchLogResponse]
    total: int
    skip: int
    limit: int


class EventLogWithDispatchCount(BaseModel):
    """帶派發計數的事件日誌響應模型"""

    id: str
    topic_id: str
    source_ip: Optional[str] = None
    headers: Optional[dict] = None
    content_type: Optional[str] = None
    payload: Optional[str] = None
    status: str
    received_at: datetime
    dispatch_count: int = 0  # 關聯的派發日誌數量

    class Config:
        from_attributes = True

    @field_validator('status', mode='before')
    @classmethod
    def status_enum_to_string(cls, v):
        """將狀態枚舉轉換為字符串"""
        if isinstance(v, EventLogStatus):
            return v.name
        return v


class EventLogWithDispatches(BaseModel):
    """帶派發記錄的事件日誌響應模型（階層式）"""

    id: str
    topic_id: str
    source_ip: Optional[str] = None
    headers: Optional[dict] = None
    content_type: Optional[str] = None
    payload: Optional[str] = None
    status: str
    received_at: datetime
    dispatch_count: int = 0
    dispatches: List[DispatchLogResponse] = []  # 關聯的派發日誌列表

    class Config:
        from_attributes = True

    @field_validator('status', mode='before')
    @classmethod
    def status_enum_to_string(cls, v):
        """將狀態枚舉轉換為字符串"""
        if isinstance(v, EventLogStatus):
            return v.name
        return v


class HierarchicalLogListResponse(BaseModel):
    """階層式日誌列表響應模型"""

    items: List[EventLogWithDispatchCount]
    total: int
    skip: int
    limit: int

File: frontend/src/types/log.ts
// 日誌相關類型定義

// 事件日誌狀態
export type EventLogStatus = 'RECEIVED' | 'QUEUED' | 'FAILED_VALIDATION'

// 派發日誌狀態
export type DispatchLogStatus = 'PENDING' | 'SUCCESS' | 'FAILED' | 'TIMEOUT'

// 事件日誌
export interface EventLog {
  id: string
  topic_id: string
  source_ip: string
  headers: Record<string, string>
  content_type: string
  payload: string
  status: EventLogStatus
  received_at: string
  processed_at?: string
}

// 派發日誌
export interface DispatchLog {
  id: string
  event_log_id: string
  subscription_id: string
  attempt: number
  status: DispatchLogStatus
  response_status_code?: number
  response_body?: string
  error_message?: string
  dispatched_at: string
  completed_at?: string
}

// 事件日誌列表響應
export interface EventLogListResponse {
  items: EventLog[]
  total: number
  skip: number
  limit: number
}

// 派發日誌列表響應
export interface DispatchLogListResponse {
  items: DispatchLog[]
  total: number
  skip: number
  limit: number
}

// 日誌服務錯誤
export interface LogError {
  message: string
  code?: string
}

// API 響應包裝器
export interface LogApiResponse<T> {
  data?: T
  error?: LogError
  loading: boolean
}

// 日誌篩選參數
export interface EventLogFilterParams {
  topic_id?: string
  status?: EventLogStatus
  source_ip?: string
  date_from?: string
  date_to?: string
  skip?: number
  limit?: number
}

export interface DispatchLogFilterParams {
  event_log_id?: string
  subscription_id?: string
  status?: DispatchLogStatus
  date_from?: string
  date_to?: string
  skip?: number
  limit?: number
}

// 日誌搜尋參數
export interface LogSearchParams {
  query?: string // 關鍵字搜尋
  type?: 'events' | 'dispatches' | 'all'
}

// 綜合日誌項目（用於統一顯示）
export interface UnifiedLogItem {
  id: string
  type: 'event' | 'dispatch'
  timestamp: string
  status: string
  topic_id?: string
  subscription_id?: string
  source_ip?: string
  response_status_code?: number
  error_message?: string
}

// 綜合日誌列表響應
export interface UnifiedLogListResponse {
  items: UnifiedLogItem[]
  total: number
  skip: number
  limit: number
}

// 日誌統計摘要
export interface LogStats {
  total_events: number
  total_dispatches: number
  success_rate: number
  average_response_time: number
  recent_errors: number
  last_updated: string
}
